src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala

```diff
@@ -53,56 +50,84 @@ trait WorkerMain[S] {
           }
         }
 
-        try {
-          @tailrec
-          def process(ctx: S): Unit = {
-            val request = WorkerProtocol.WorkRequest.parseDelimitedFrom(stdin)
-            if (request == null) {
-              return
-            }
-            val args = request.getArgumentsList.toArray(Array.empty[String])
+        @tailrec
+        def process(ctx: S): Unit = {
+          val request = WorkerProtocol.WorkRequest.parseDelimitedFrom(stdin)
+          if (request == null) {
+            return
+          }
+          val args = request.getArgumentsList.toArray(Array.empty[String])
 
-            val outStream = new ByteArrayOutputStream
-            val out = new PrintStream(outStream)
-            val requestId = request.getRequestId()
-            System.out.println(s"WorkRequest $requestId received with args: ${request.getArgumentsList}")
+          // We go through this hullabaloo with output streams being defined out here, so we can
+          // close them after the async work in the Future is all done.
+          // If we do something synchronous with Using, then there's a race condition where the
+          // streams can get closed before the Future is completed.
+          var outStream: ByteArrayOutputStream = null
+          var out: PrintStream = null
```

Comment:
Why do these have to be set to null first?

Reply:
Because if I create the stream here and something goes wrong, then we could leak memory by not closing the stream. By doing so in the future, I'm trying to avoid that. It needs to be declared outside the future, so I can reference it in a subsequent future to be able to close it after the original future has done its work. That's what I'm trying to do with the andThen calls.